core/src/main/java/org/apache/accumulo/core/metadata/TabletFile.java

```diff
@@ -104,21 +110,50 @@ public String getFileName() {
   }
 
   /**
-   * Exact string that is stored in the metadata table
+   * Return a string for reading the tablet file. Doesn't have to be exact string in metadata.
    */
-  public String getMetadataEntry() {
-    return metadataEntry;
+  public String getMetaRead() {
+    return normalizedPath;
   }
 
   /**
-   * Exact string that is stored in the metadata table but as a Text object
+   * Return a string for inserting a new tablet file.
    */
-  public Text getMetadataText() {
-    return new Text(metadataEntry);
+  public String getMetaInsert() {
+    return normalizedPath;
   }
 
-  public String getNormalizedPath() {
-    return normalizedPath;
+  /**
+   * Return a new Text object of {@link #getMetaInsert()}
+   */
+  public Text getMetaInsertText() {
+    return new Text(getMetaInsert());
+  }
+
+  /**
+   * New file was written to metadata so update the entry.
+   */
+  public void inserted() {
+    this.metadataEntry = Optional.of(normalizedPath);
+  }
+
+  /**
+   * Return the exact string that is stored in the metadata table. This is important for updating
+   * and deleting metadata entries. If the exact string is not used, erroneous entries can pollute
+   * the metadata table.
+   */
+  public String getMetaUpdateDelete() {
+    if (metadataEntry.isEmpty())
+      throw new IllegalStateException(
+          "TabletFile " + metaPath + " does not have a metadata entry.");
+    return metadataEntry.get();
```

Comment:
This seems to be protecting against programming errors, where the TabletFile on which this method is being called was not constructed from a metadata entry. However, a better way to protect against these kinds of programming errors is to put this method on a subclass. So, by the type hierarchy, you enforce the fact that this method can *only* ever be called on a subclass, which is a type where construction requires a metadata entry.

Reply:
This is a good idea too, and may work well with @keith-turner 's suggestion.  I will update the PR with the other minor suggestions before attempting this change.